crates/katana/core/src/messaging/mod.rs

```diff
@@ -0,0 +1,166 @@
+    async fn gather_messages(
+        &self,
+        from_block: u64,
+        max_blocks: u64,
+    ) -> MessengerResult<(u64, Vec<L1HandlerTransaction>)>;
+
+    /// Computes the hash of the given messages and sends them to the settlement chain.
+    ///
+    /// Once message's hash is settled, one must send a transaction (with the message content)
+    /// on the settlement chain to actually consume it.
+    ///
+    /// # Arguments
+    ///
+    /// * `messages` - Messages to settle.
+    async fn settle_messages(&self, messages: &[MsgToL1]) -> MessengerResult<Vec<String>>;
+}
+
+pub enum AnyMessenger {
+    Ethereum(Arc<AsyncRwLock<EthereumMessenger>>),
+    Starknet(Arc<AsyncRwLock<StarknetMessenger>>),
+}
+
+impl AnyMessenger {
+    pub async fn from_file(file_path: &PathBuf) -> MessengerResult<Self> {
+        let mut file = File::open(file_path).expect("Messaging config file error");
+        let mut json_string = String::new();
+        file.read_to_string(&mut json_string).expect("Messaging config file read error");
+
+        let config: MessagingConfig =
+            serde_json::from_str(&json_string).expect("Messaging config file parsing error");
+
+        Self::from_config(config).await
+    }
+
+    pub async fn from_config(config: MessagingConfig) -> MessengerResult<Self> {
+        if config.contract_address.len() < 50 {
+            match EthereumMessenger::new(config.clone()).await {
+                Ok(m_eth) => {
+                    info!(MSGING_TARGET, "Messaging enabled [Ethereum]");
+                    Ok(AnyMessenger::Ethereum(m_eth))
+                }
+                Err(e_eth) => {
+                    error!("Ethereum messenger init failed: {:?}", e_eth);
+                    Err(MessengerError::InitError)
+                }
+            }
+        } else {
+            match StarknetMessenger::new(config.clone()).await {
+                Ok(m_sn) => {
+                    info!(target: MSGING_TARGET, "Messaging enabled [Starknet]");
+                    Ok(AnyMessenger::Starknet(m_sn))
+                }
+                Err(e_sn) => {
+                    error!(target: MSGING_TARGET,
+                           "Starknet messenger init failed: {:?}", e_sn);
+                    Err(MessengerError::InitError)
+                }
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl Messenger for AnyMessenger {
+    async fn gather_messages(
+        &self,
+        from_block: u64,
+        max_blocks: u64,
+    ) -> MessengerResult<(u64, Vec<L1HandlerTransaction>)> {
+        match self {
+            Self::Ethereum(inner) => {
+                inner.read().await.gather_messages(from_block, max_blocks).await
+            }
+            Self::Starknet(inner) => {
+                inner.read().await.gather_messages(from_block, max_blocks).await
+            }
+        }
+    }
+
+    async fn settle_messages(&self, messages: &[MsgToL1]) -> MessengerResult<Vec<String>> {
```

Comment:
why is it returning `Vec<String>`? i assume to accomodate for both `Ethereum` and `Starknet` hash type?

Reply:
Yes, it was the first reason to avoid exposing two different types in enum variants. But it's perhaps more rust-way to have each respective type in an enum.
I'll make a proposition for this one to see if it's looks better.